### PR TITLE
Support specifying executor for accessing manifests

### LINKED ISSUE #3776: Flink connector thrown `Trying to access closed classloader`
Hi, i'm trying to test the unreleased Flink 1.13 connector on a Flink 1.13.1 cluster running on k8s.

The task is simple like `INSERT OVERWRITE a SELECT * FROM b`.

When i submit a task for the first time, it's ok.
But when i submit the task for the second time(run on the same TaskManager with the first time), the following exception is thrown, it's weird.

<details>
<summary>java.lang.IllegalStateException: Trying to access closed classloader</summary>

```
2021-12-20 11:53:11,697 INFO  org.apache.flink.runtime.taskmanager.Task                    [] - IcebergFilesCommitter -> Sink: IcebergSink ice.t1.d (1/1)#0 (e74babbac1ebb51fe72958bd177ae534) switched from INITIALIZING to RUNNING.
2021-12-20 11:53:11,854 INFO  org.apache.iceberg.flink.sink.IcebergFilesCommitter          [] - Committing dynamic partition overwrite with 1 data files and 0 delete files to table ice.t1.d
2021-12-20 11:53:11,946 WARN  org.apache.flink.runtime.taskmanager.Task                    [] - IcebergFilesCommitter -> Sink: IcebergSink ice.t1.d (1/1)#0 (e74babbac1ebb51fe72958bd177ae534) switched from RUNNING to FAILED with failure cause: java.lang.IllegalStateException: Trying to access closed classloader. Please check if you store classloaders directly or indirectly in static fields. If the stacktrace suggests that the leak occurs in a third party library and cannot be fixed immediately, you can disable this check with the configuration 'classloader.check-leaked-classloader'.
	at org.apache.flink.runtime.execution.librarycache.FlinkUserCodeClassLoaders$SafetyNetWrapperClassLoader.ensureInner(FlinkUserCodeClassLoaders.java:164)
	at org.apache.flink.runtime.execution.librarycache.FlinkUserCodeClassLoaders$SafetyNetWrapperClassLoader.loadClass(FlinkUserCodeClassLoaders.java:172)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Unknown Source)
	at org.apache.iceberg.common.DynClasses$Builder.impl(DynClasses.java:68)
	at org.apache.iceberg.avro.GenericAvroReader$ReadBuilder.record(GenericAvroReader.java:85)
	at org.apache.iceberg.avro.GenericAvroReader$ReadBuilder.record(GenericAvroReader.java:72)
	at org.apache.iceberg.avro.AvroSchemaVisitor.visit(AvroSchemaVisitor.java:50)
	at org.apache.iceberg.avro.GenericAvroReader.initReader(GenericAvroReader.java:47)
	at org.apache.iceberg.avro.GenericAvroReader.setSchema(GenericAvroReader.java:53)
	at org.apache.iceberg.avro.ProjectionDatumReader.newDatumReader(ProjectionDatumReader.java:80)
	at org.apache.iceberg.avro.ProjectionDatumReader.setSchema(ProjectionDatumReader.java:69)
	at org.apache.iceberg.shaded.org.apache.avro.file.DataFileStream.initialize(DataFileStream.java:133)
	at org.apache.iceberg.shaded.org.apache.avro.file.DataFileReader.<init>(DataFileReader.java:130)
	at org.apache.iceberg.shaded.org.apache.avro.file.DataFileReader.<init>(DataFileReader.java:122)
	at org.apache.iceberg.shaded.org.apache.avro.file.DataFileReader.openReader(DataFileReader.java:66)
	at org.apache.iceberg.avro.AvroIterable.newFileReader(AvroIterable.java:100)
	at org.apache.iceberg.avro.AvroIterable.getMetadata(AvroIterable.java:66)
	at org.apache.iceberg.ManifestReader.<init>(ManifestReader.java:103)
	at org.apache.iceberg.ManifestFiles.read(ManifestFiles.java:87)
	at org.apache.iceberg.SnapshotProducer.newManifestReader(SnapshotProducer.java:378)
	at org.apache.iceberg.MergingSnapshotProducer$DataFileFilterManager.newManifestReader(MergingSnapshotProducer.java:716)
	at org.apache.iceberg.ManifestFilterManager.filterManifest(ManifestFilterManager.java:295)
	at org.apache.iceberg.ManifestFilterManager.lambda$filterManifests$0(ManifestFilterManager.java:186)
	at org.apache.iceberg.util.Tasks$Builder.runTaskWithRetry(Tasks.java:405)
	at org.apache.iceberg.util.Tasks$Builder.access$300(Tasks.java:71)
	at org.apache.iceberg.util.Tasks$Builder$1.run(Tasks.java:311)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
```
</details>

### PR DESCRIPTION
**Pull request by yittg**
Fixes #3776

### CODE REVIEW

**Comment by yittg** [comment]
@rdblue i try to provide a way to use a particular executor for accessing manifests, with which we can avoid sharing executor btw flink jobs.
Do you think it is reasonable, and have a look?

File: core/src/main/java/org/apache/iceberg/CatalogSupport.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ */
+
+package org.apache.iceberg;
+
+import java.util.concurrent.ExecutorService;
+
+public interface CatalogSupport {
+
+  void withManifestAccessExecutor(ExecutorService executor);
+}
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/CatalogSupport.java:22]: This shouldn't be passed through the catalog or table operations. Instead, it should be passed through the scan interfaces by the processing engine. I think you probably want to add this to `TableScan` instead.

Comment (yittg) [reply core/src/main/java/org/apache/iceberg/CatalogSupport.java]: Sounds good, i've updated the commits, PTAL.

**Comment by yittg** [comment]
@rdblue Please review this PR when you are free.

File: api/src/main/java/org/apache/iceberg/ExpireSnapshots.java [hunk 2]
```diff
@@ -98,6 +98,15 @@
...
   ExpireSnapshots executeDeleteWith(ExecutorService executorService);
 
+  /**
+   * Passes an alternative executor service that will be used for accessing manifests.
+   * If this method is not called, the default worker pool will be used.
+   *
+   * @param executorService an executor service to access manifests
+   * @return this for method chaining
+   */
+  ExpireSnapshots accessManifestsWith(ExecutorService executorService);
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/ExpireSnapshots.java:108]: The methods that run this are named `plan`: `planFiles` and `planSplits` so I'd probably change this to `planWith`. Is that okay with you, @yittg?

File: api/src/main/java/org/apache/iceberg/TableScan.java [hunk 3]
```diff
@@ -159,6 +160,15 @@ default TableScan select(String... columns) {
...
   TableScan appendsAfter(long fromSnapshotId);
 
+  /**
+   * Create a new {@link TableScan} to use a particular executor to access manifests. The default worker pool will be
+   * used by default.
+   *
+   * @param executorService the provided executor.
+   * @return a table scan which use the provided executor to access manifests.
+   */
+  TableScan accessManifestsWith(ExecutorService executorService);
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/TableScan.java:170]: `planWith`?

File: api/src/main/java/org/apache/iceberg/SnapshotUpdate.java [hunk 4]
```diff
@@ -52,4 +53,13 @@
    */
   ThisT stageOnly();
 
+  /**
+   * Use a particular executor to access manifests. The default worker pool will be used by default.
+   *
+   * @param executorService the provided executor.
+   * @return this for method chaining.
+   */
+  default ThisT accessManifestsWith(ExecutorService executorService) {
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/SnapshotUpdate.java:62]: Could you separate this into a PR for the scan updates and a PR for some of the operation updates (like ExpireSnapshots)? This one together is really large.

**Comment by yittg** [comment]
Hi @rdblue, i've created a new issue #4145 to track the api changes for different types of operations.
Please switch to the new PRs, after these PRs are merged, we can continue to fix #3776 in Flink.

**Comment by rdblue** [comment]
I'm going to close this because it is superceded by #4145 and its subtasks. Thanks for working on this, @yittg, this is really valuable!

**Pull request closed without merging**